src/components/registration-form.js

```diff
@@ -1,148 +1,285 @@
-  onRegistrationStarted,
-  onRegistrationSucceeded
-}) {
-  const [form] = Form.useForm();
+const STEP = {
+  enterData: 'enterData',
+  enterCode: 'enterCode',
+  completed: 'completed'
+};
+
+const STEP_STATE = {
+  wait: 'wait',
+  process: 'process',
+  finish: 'finish',
+  error: 'error'
+};
+
+const createDefaultSteps = () => [
+  {
+    key: STEP.enterData,
+    status: STEP_STATE.wait,
+    icon: <UserOutlined />,
+    translationKey: 'step_enterData'
+  },
+  {
+    key: STEP.enterCode,
+    status: STEP_STATE.wait,
+    icon: <SolutionOutlined />,
+    translationKey: 'step_enterCode'
+  },
+  {
+    key: STEP.completed,
+    status: STEP_STATE.wait,
+    icon: <SmileOutlined />,
+    translationKey: 'step_completed'
+  }
+];
+
+const setStepError = (steps, key) => steps.map(step => step.key === key ? { ...step, status: STEP_STATE.error } : step);
+const setStepSuccess = (steps, key) => steps.map(step => step.key === key ? { ...step, status: STEP_STATE.finish } : step);
+
+function RegistrationForm() {
+  const setUser = useSetUser();
   const settings = useSettings();
   const { uiLanguage } = useLocale();
-  const [user, setUser] = useState(null);
+  const [enterDataForm] = Form.useForm();
+  const [enterCodeForm] = Form.useForm();
+  const { formatDuration } = useDateFormat();
   const userApiClient = useService(UserApiClient);
   const { t } = useTranslation('registrationForm');
+  const [isLoading, setIsLoading] = useState(false);
+  const [createdUser, setCreatedUser] = useState(null);
+  const [steps, setSteps] = useState(createDefaultSteps());
   const [forbiddenEmails, setForbiddenEmails] = useState([]);
+  const [currentStepKey, setCurrentStepKey] = useState(STEP.enterData);
+  const [invalidVerificationCodes, setInvalidVerificationCodes] = useState([]);
 
-  if (formRef) {
-    formRef.current = form;
-  }
+  useEffect(() => {
+    if (currentStepKey === STEP.enterData) {
+      setCreatedUser(null);
+      setForbiddenEmails([]);
+      setInvalidVerificationCodes([]);
+    }
+    setSteps(previousSteps => {
+      const newSteps = currentStepKey === STEP.enterData ? createDefaultSteps() : previousSteps.map(step => ({ ...step }));
+      for (const step of newSteps) {
+        if (step.key === currentStepKey) {
+          step.status = STEP_STATE.process;
+        }
+      }
+      return newSteps;
+    });
+  }, [currentStepKey]);
+
+  const handleEnterDataButtonClick = () => {
+    enterDataForm.submit();
+  };
+
+  const handleEnterCodeButtonClick = () => {
+    enterCodeForm.submit();
+  };
 
-  const register = async ({ email, password, displayName }) => {
+  const handleEnterDataFinish = async ({ email, password, displayName }) => {
     try {
-      onRegistrationStarted();
-      const { result, user: registeredUser } = await userApiClient.register({ email, password, displayName });
+      setIsLoading(true);
+      const { result, user } = await userApiClient.requestRegistration({
+        email,
```

Comment:
Should we trim this guy too? 🤔 We do it at login but not here. Or we could maybe trim both on the server side actually, might be better as we lowercase the email there anyway.

Reply:
I trimmed it here as I consider it more a UI concern than the lowercasing, so the API gets already sanitized values.